{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4211ea30-9d2d-4a8c-8da5-858e98661b00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: zfit-physics in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (0.6.1)\n",
      "Requirement already satisfied: zfit>=0.12 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit-physics) (0.16.0)\n",
      "Requirement already satisfied: attrs in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (21.4.0)\n",
      "Requirement already satisfied: boost-histogram in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (1.4.0)\n",
      "Requirement already satisfied: colorama in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (0.4.6)\n",
      "Requirement already satisfied: colored in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (1.4.4)\n",
      "Requirement already satisfied: colorlog in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (6.7.0)\n",
      "Requirement already satisfied: deprecated in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (1.2.14)\n",
      "Requirement already satisfied: dill in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (0.3.7)\n",
      "Requirement already satisfied: dotmap in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (1.3.30)\n",
      "Requirement already satisfied: frozendict in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (2.3.8)\n",
      "Requirement already satisfied: hist in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (2.7.2)\n",
      "Requirement already satisfied: iminuit>=2.3 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (2.24.0)\n",
      "Requirement already satisfied: jacobi in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (0.9.2)\n",
      "Requirement already satisfied: numdifftools in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (0.9.41)\n",
      "Requirement already satisfied: numpy>=1.16 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (1.22.4)\n",
      "Requirement already satisfied: ordered-set in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (4.1.0)\n",
      "Requirement already satisfied: pandas in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (1.3.5)\n",
      "Requirement already satisfied: pydantic<2.0.0 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (1.10.12)\n",
      "Requirement already satisfied: pyyaml in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (6.0)\n",
      "Requirement already satisfied: scipy>=1.2 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (1.7.3)\n",
      "Requirement already satisfied: tabulate in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (0.8.9)\n",
      "Requirement already satisfied: tensorflow<2.14,>=2.13 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (2.13.0)\n",
      "Requirement already satisfied: tensorflow-probability<0.22,>=0.20 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (0.21.0)\n",
      "Requirement already satisfied: texttable in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (1.6.7)\n",
      "Requirement already satisfied: uhi in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (0.3.3)\n",
      "Requirement already satisfied: uproot>=4 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (4.1.9)\n",
      "Requirement already satisfied: xxhash in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (3.3.0)\n",
      "Requirement already satisfied: zfit-interface in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (0.0.3)\n",
      "Requirement already satisfied: typing-extensions in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from zfit>=0.12->zfit-physics) (4.5.0)\n",
      "Requirement already satisfied: absl-py>=1.0.0 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (2.0.0)\n",
      "Requirement already satisfied: astunparse>=1.6.0 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (1.6.3)\n",
      "Requirement already satisfied: flatbuffers>=23.1.21 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (23.5.26)\n",
      "Requirement already satisfied: gast<=0.4.0,>=0.2.1 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (0.3.3)\n",
      "Requirement already satisfied: google-pasta>=0.1.1 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (0.2.0)\n",
      "Requirement already satisfied: grpcio<2.0,>=1.24.3 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (1.58.0)\n",
      "Requirement already satisfied: h5py>=2.9.0 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (2.10.0)\n",
      "Requirement already satisfied: keras<2.14,>=2.13.1 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (2.13.1)\n",
      "Requirement already satisfied: libclang>=13.0.0 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (16.0.6)\n",
      "Requirement already satisfied: opt-einsum>=2.3.2 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (3.3.0)\n",
      "Requirement already satisfied: packaging in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (21.3)\n",
      "Requirement already satisfied: protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (4.24.3)\n",
      "Requirement already satisfied: setuptools in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (60.5.0)\n",
      "Requirement already satisfied: six>=1.12.0 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (1.15.0)\n",
      "Requirement already satisfied: tensorboard<2.14,>=2.13 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (2.13.0)\n",
      "Requirement already satisfied: tensorflow-estimator<2.14,>=2.13.0 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (2.13.0)\n",
      "Requirement already satisfied: termcolor>=1.1.0 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (1.1.0)\n",
      "Requirement already satisfied: wrapt>=1.11.0 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (1.12.1)\n",
      "Requirement already satisfied: tensorflow-io-gcs-filesystem>=0.23.1 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (0.34.0)\n",
      "Requirement already satisfied: decorator in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorflow-probability<0.22,>=0.20->zfit>=0.12->zfit-physics) (5.1.1)\n",
      "Requirement already satisfied: cloudpickle>=1.3 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorflow-probability<0.22,>=0.20->zfit>=0.12->zfit-physics) (2.0.0)\n",
      "Requirement already satisfied: dm-tree in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorflow-probability<0.22,>=0.20->zfit>=0.12->zfit-physics) (0.1.6)\n",
      "Requirement already satisfied: histoprint>=2.2.0 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from hist->zfit>=0.12->zfit-physics) (2.4.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from pandas->zfit>=0.12->zfit-physics) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2017.3 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from pandas->zfit>=0.12->zfit-physics) (2021.3)\n",
      "Requirement already satisfied: wheel<1.0,>=0.23.0 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from astunparse>=1.6.0->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (0.37.1)\n",
      "Requirement already satisfied: click>=7.0.0 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from histoprint>=2.2.0->hist->zfit>=0.12->zfit-physics) (8.1.7)\n",
      "Requirement already satisfied: google-auth<3,>=1.6.3 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (2.23.0)\n",
      "Requirement already satisfied: google-auth-oauthlib<1.1,>=0.5 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (1.0.0)\n",
      "Requirement already satisfied: markdown>=2.6.8 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (3.3.6)\n",
      "Requirement already satisfied: requests<3,>=2.21.0 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (2.27.1)\n",
      "Requirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in /afs/desy.de/user/j/jipatel/.local/lib/python3.8/site-packages (from tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (0.7.1)\n",
      "Requirement already satisfied: werkzeug>=1.0.1 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (2.0.2)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from packaging->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (3.0.6)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (4.2.4)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (0.2.8)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (4.8)\n",
      "Requirement already satisfied: urllib3<2.0 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (1.26.8)\n",
      "Requirement already satisfied: requests-oauthlib>=0.7.0 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from google-auth-oauthlib<1.1,>=0.5->tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (1.3.0)\n",
      "Requirement already satisfied: importlib-metadata>=4.4 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from markdown>=2.6.8->tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (4.10.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from requests<3,>=2.21.0->tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (2021.10.8)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from requests<3,>=2.21.0->tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (2.0.10)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from requests<3,>=2.21.0->tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (3.3)\n",
      "Requirement already satisfied: zipp>=0.5 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from importlib-metadata>=4.4->markdown>=2.6.8->tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (3.7.0)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from pyasn1-modules>=0.2.1->google-auth<3,>=1.6.3->tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (0.4.8)\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in /cvmfs/belle.cern.ch/el7/externals/v02-00-02/Linux_x86_64/common/lib/python3.8/site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<1.1,>=0.5->tensorboard<2.14,>=2.13->tensorflow<2.14,>=2.13->zfit>=0.12->zfit-physics) (3.1.1)\n",
      "DEPRECATION: pytorch-lightning 1.5.6 has a non-standard dependency specifier torch>=1.7.*. pip 24.0 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pytorch-lightning or contact the author to suggest that they release a version with a conforming dependency specifiers. Discussion can be found at https://github.com/pypa/pip/issues/12063\n",
      "\n",
      "[notice] A new release of pip is available: 23.3.2 -> 24.0\n",
      "[notice] To update, run: pip install --upgrade pip\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install zfit-physics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f7728d26-eaf8-45a0-a04f-f96fe7ac9fe5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import uproot\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import crystalball\n",
    "from scipy.optimize import curve_fit,Bounds\n",
    "import zfit_physics as zphysics\n",
    "import zfit\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "import matplotlib.pyplot as plt\n",
    "import mplhep\n",
    "import uncertainties as unc\n",
    "plt.style.use('belle2')\n",
    "#%jsroot on"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e57a206c-397e-44f0-a8c3-a8a2886f8edd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Belle2 (light-2311-nebelung)",
   "language": "python",
   "name": "belle2_light-2311-nebelung"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
